### ISSUE TITLE
Reading progress window - not full height

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#11177]
* KOReader version: v2023.10-23
* Device: Kindle Scribe (5.16.2)

#### Issue and Steps to reproduce
Top menu - Reading statisctics - Reading progress. The window is not full height - I can see part of the currently open book and the progress bar at the bottom.

![crash_reading_progress_2023-11-26_225406](https://github.com/koreader/koreader/assets/6279855/ff03c790-2cfd-48a4-84bc-0f59a069d357)

##### `crash.log` (if applicable)
Clean logs with opening KOReader and going straight to top menu - Reading statisctics - Reading progress.
[crash_reading_progress.log.txt](https://github.com/koreader/koreader/files/13493622/crash_reading_progress.log.txt)

**Comment by hius07** [comment]
(Cannot dig deeper at the moment)
What if you change `self.height` to `self.screen_height` in
https://github.com/koreader/koreader/blob/dce12de09d4c4030954ce00e72f3986c89a2a932/plugins/statistics.koplugin/readerprogress.lua#L72

**Comment by NiLuJe** [comment]
That sounds like a winner, none of the callers actually pass a height/width to the constructor ;).

**Comment by mergen3107** [comment]
Nope, that didn't do. Still the same issue

**Comment by mergen3107** [comment]
https://github.com/koreader/koreader/blob/dce12de09d4c4030954ce00e72f3986c89a2a932/plugins/statistics.koplugin/readerprogress.lua#L277

**Comment by NiLuJe** [comment]
I can't reproduce this anywhere (the Elipsa isn't 300dpi like the Scribe), so this might get tricky ;).

On the off-chance it's yet another rotation quirk: does this happen when you start KOReader with the device properly upright?

**Comment by mergen3107** [comment]
Interesting, OK! I'll keep looking. 

Dpi-wise, 227/300 is 25% apart. This case, on the contrary, is 5% off (120 px / 2580 px). 

Yes, happens in both `U` and `D`, and I didn't find it anywhere else yet, so laser focusing this plugin for now.

For example, books reading stats (with minutes per page and stuff), systems stats, page browser, book map and many other full screen windows are OK. 

I'll also check Cloud, and full screen dictionary later.

**Comment by mergen3107** [comment]
Thanks!

**Comment by mergen3107** [comment]
FWIW, Android 10 (Onyx Boox Kon-Tiki 2) also has not full height, but Kobo Clara HD doesn't. Didn't check their KOReader versions though, didn't have enough time.

**Comment by mergen3107** [comment]
Onyx Boox Kon-Tiki 2: vertical 1872 px, 300 dpi

**Comment by mergen3107** [comment]
Added `logger` stuff:
```
12/02/23-22:19:13 DEBUG self.screen_height =  2480 
12/02/23-22:19:13 DEBUG self.screen_width =  1860 
```
Correct!

I will check now what the bottom `Session` and `Today` groups are doing, maybe they are shifting up?

**Comment by mergen3107** [comment]
(let me get a new brach and add the changes there to show you. Wanna make it a habit for easier tracking...)

**Comment by mergen3107** [comment]
Here is the full log of the last opening KOReader (restarting), and then opening Top menu - Reading Statistics - Reading Progress.
[crash_readerprogress_debug.txt](https://github.com/koreader/koreader/files/13537992/crash_readerprogress_debug.txt)

**Comment by mergen3107** [comment]
Alright, I think I found the two culprits!

https://github.com/koreader/koreader/blob/b361cec4ff97de18d4a5becc63884f586dd7e626/plugins/statistics.koplugin/readerprogress.lua#L39-L47

These two paddings!

The thing is, they just _happened_ to be large enough on other devices to be able to cover beyond the bottom edge of the screen. They are in pixels.

I will try to make it so each padding is calculated based on the actual screen size and content.

Content is fixed vertical-wise:
Three headers (both single and double), seven days, two horizontal groups, paddings between horizontal groups, paddings between days.

**Comment by Frenzie** [comment]
>  They are in pixels.

There'll be a scaleBySize on them.

**Comment by Frenzie** [comment]
Yes, but you'll get something like 25 * 1.8 = 45 for example.

So in what you wrote:

> I will try to make it so each padding is calculated based on the actual screen size and content.

It is based on the actual screen size already.

**Comment by mergen3107** [comment]
@Frenzie 
Thanks! I am still a little confused with how this function works, scaleBySize. Where does 1.8 come from?

Where can I see this function definition itself?

**Comment by mergen3107** [comment]
> and use that to scale all our abstract sizes
Thanks!
I think I understand dpi conversion on different physical dpi and user-selected dpi.

However, I also think the problem with the ReadingProgress window is that it is "blind".
It just assumes some size/dpi, then just "stack" horizontal groups/headers from top and bottom.
I think a better way to draw this is to draw the top group `Last week` relative to the top, but, for example, `Session` and `Today` group relative to the very bottom of the screen size. Then `Week progress` in between those (maybe relative to the bottom group, with some conservative value of `span` in order to to overflow to the top group).

**Comment by poire-z** [comment]
You're right.
This widget is some very old one build by somebody who didn't really understand our widget stuff (it took me some time to get there :)). And nobody really cared about it to rewrite it.
The proper way is to go from the available height, split it in shares, see what to put in each share, and then compute the font size to use for stuff to fit in that share. (And not going from some fixed abstract font size up to their container up to the screen, with either some blank space or some overflow).

**Comment by mergen3107** [comment]
This is how it looks like:
![57_0 6x65_Reader_2023-12-03_122907](https://github.com/koreader/koreader/assets/6279855/0df46ddd-0ede-4b91-b136-7f4c3f7dbaf1)

**Comment by mergen3107** [comment]
This is how it looks at different DPI settings. Weird things is that after I tried different DPI settings, from now on when I take screenshot (either in KOReader two opposite corners or via SSH `screenshot` command), the screenshot does not have the progress bar O_o. I see it on any other settings than `DPI = Auto`. I can't even get the progress bar again in the same `DPI = 120` as in the OP.

`DPI = Auto` (cannot see progress bar)
![DPI__AUTO_Reader_2023-12-03_163815](https://github.com/koreader/koreader/assets/6279855/83cc406b-195f-4540-a032-b9dde885cd31)

`DPI = 300` (can see progress bar - but it is NOT on the screenshots)
![DPI_300_Reader_2023-12-03_163926](https://github.com/koreader/koreader/assets/6279855/d9d6219b-1480-44e4-a76c-17b8c5f45ca3)

`DPI = 120` (can see progress bar - but it is NOT on the screenshots anymore)
![DPI_120_Reader_2023-12-03_163849](https://github.com/koreader/koreader/assets/6279855/d6cb6815-d1f2-4b1d-9020-bb0ee13dcc86)

**Comment by Frenzie** [comment]
That's with a restart? Things like language and DPI could potentially be a little weird without.

**Comment by NiLuJe** [comment]
> Maybe, but how did I get the progress bar in the screenshots before? O_o

Because of what @hius07 mentioned first, probably. The FrameContainer wasn't painting to the whole screen; but now it is,n and what's missing is just the refresh region being too small, as you noticed in https://github.com/koreader/koreader/issues/11177#issuecomment-1837342333

**Comment by mergen3107** [comment]
Still weird! 
Why is it so inconsistent with itself? Isn't this a problem?

**Comment by NiLuJe** [comment]
Because https://github.com/koreader/koreader/issues/11177#issuecomment-1837542416 I guess ;).

**Comment by mergen3107** [comment]
Closed by #11226.
